spec/services/course/material/zip_download_service.rb

```diff
@@ -3,14 +3,110 @@
+          folder_c_path = subject + '/' + folder_c.name
+          folder_d_path = folder_c_path + '/' + folder_d.name
+          expect(File).to exist(folder_b_path)
+          expect(File).to exist(folder_c_path)
+          expect(File).to exist(folder_d_path)
+
+          material_a_path = subject + '/' + material_a.name
+          material_b_path = folder_b_path + '/' + material_b.name
+          material_d_path = folder_d_path + '/' + material_d.name
+          expect(File).to exist(material_a_path)
+          expect(File).to exist(material_b_path)
+          expect(File).to exist(material_d_path)
+
+          # File contents should be the same
+          expect(
+            FileUtils.compare_file(
+              material_a_path,
+              material_a.attachment.file_upload.path
+            )
+          ).to be_truthy
+
+          expect(
+            FileUtils.compare_file(
+              material_b_path,
+              material_b.attachment.file_upload.path
+            )
+          ).to be_truthy
+
+          expect(
+            FileUtils.compare_file(
+              material_d_path,
+              material_d.attachment.file_upload.path
+            )
+          ).to be_truthy
+        end
+      end
+
+      context 'when some of the materials are selected' do
+        let(:materials) { [material_a, material_b] }
+
+        it 'only downloads the selected materials' do
+          expect(subject).to be_present
+
+          # Only selected files and their parent folders should be downloaded
+          folder_b_path = subject + '/' + folder_b.name
+          folder_c_path = subject + '/' + folder_c.name
+          expect(File).to exist(folder_b_path)
+          expect(File).not_to exist(folder_c_path)
+
+          material_a_path = subject + '/' + material_a.name
+          material_b_path = folder_b_path + '/' + material_b.name
+          expect(File).to exist(material_a_path)
+          expect(File).to exist(material_b_path)
+        end
+      end
+    end
+
+    describe '.zip_dir' do
+      let(:dir) { Course::Material::ZipDownloadService.send(:download, materials, folder_a) }
+      subject { Course::Material::ZipDownloadService.send(:zip_dir, dir) }
+
+      it 'zips the directory' do
         expect(subject).to be_present
+
+        file_names = Zip::File.open(subject) { |f| f.map(&:name) }
+        expect(file_names).to include(folder_b.name + '/')
+        expect(file_names).to include(folder_c.name + '/')
+        expect(file_names).to include(material_a.name)
+        expect(file_names).to include(folder_b.name + '/' + material_b.name)
+        expect(file_names).to include(folder_c.name + '/' + folder_d.name + '/' + material_d.name)
+      end
+    end
+
+    describe '.download_and_zip' do
+      subject { Course::Material::ZipDownloadService.download_and_zip(materials, folder_a) }
+
+      it 'downloads and zips the folder' do
+        expect(File).to exist(subject)
```

Comment:
I rather `expect(File.exist?(subject)).to be_truthy`, this is a bit... strange haha